### ISSUE TITLE
sakshamguruji - User Able To Request A Withdrawal Even If The Withdrawal Queue Has Insufficient Funds To Settle

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-02-rio-network-core-protocol-judging#250]
sakshamguruji

medium

# User Able To Request A Withdrawal Even If The Withdrawal Queue Has Insufficient Funds To Settle

## Summary

A user can request a withdrawal in the RioLRTCoordinator contract , To ensure the shares owed for withdrawal don't go beyond the available shares , this check is used -> `if (sharesOwed > availableShares - withdrawalQueue().getSharesOwedInCurrentEpoch(asset)) { revert`
  
But as we will discuss , this is wrong.

## Vulnerability Detail

1.) A user wants to request a withdrawal , he calls the function `requestWithdrawal` here https://github.com/sherlock-audit/2024-02-rio-network-core-protocol/blob/main/rio-sherlock-audit/contracts/restaking/RioLRTCoordinator.sol#L99
  
  2.) The shares owed to the user is calculated to queue the withdrawal (and after the epoch is settled , user would be able to claim).
  
  3.) At L111 , available shares are calculated  , and at L112 we make the comparison
  
  ```solidity
  if (sharesOwed > availableShares - withdrawalQueue().getSharesOwedInCurrentEpoch(asset)) {
            revert INSUFFICIENT_SHARES_FOR_WITHDRAWAL();
        }
  ```
  
  This means that , the total shares that are actually avaiable for withdrawals are the total available minus the shares owed
  for withdrawals in the "current deposit" since getSharesOwedInCurrentEpoch() tells the shares owed in current deposit.
  
  4.) Since these shares are from this current epoch only i.e. I can be a user who has not yet claimed(or requested withdrawal in some old epoch) his withdraw (for some previous epoch say x-1) , but the formula
  availableShares - withdrawalQueue().getSharesOwedInCurrentEpoch(asset)) will only deduct the current epoch pending withdrawals i.e. for epoch x . 
  
  So lets's say for current epoch getSharesOwedInCurrentEpoch is 10 . 
  availableShares is 20.
  I am a user not yet claimed his withdraw for a past epoch  , let's say my withdraw was 7.

  The formula checks sharesOwed > 20 - 10 , while it should -> 20 - 10 -7
  
  5.) Due to this , withdrawals will queued even if the withdrawal queue can't handle these requests and claimWithdraw would keep reverting in such cases . 
 In short the invariant `sharesOwed > availableShares - withdrawalQueue().getSharesOwedInCurrentEpoch(asset))` is incorrect and the system would be in an unexpected state. A user could request more than what's actually available.

## Impact

  Withdrawals would be queued even when there no funds to back up the withdrawals , the invariant explained above would be incorrect.

## Code Snippet

https://github.com/sherlock-audit/2024-02-rio-network-core-protocol/blob/main/rio-sherlock-audit/contracts/restaking/RioLRTCoordinator.sol#L112

## Tool used

Manual Review

## Recommendation

 Track all the pending withdrawals across all epochs and subtract those from the available shares to get the net shares available to queue a withdrawal.

Duplicate of #361
